src/main/java/com/google/devtools/build/lib/bazel/bzlmod/VendorManager.java

```diff
@@ -57,25 +57,36 @@ public void vendorRepos(Path externalRepoRoot, ImmutableList<RepositoryName> rep
     }
 
     for (RepositoryName repo : reposToVendor) {
-      // Only re-vendor the repository if it is not up-to-date.
-      if (!isRepoUpToDate(repo, externalRepoRoot)) {
-        Path markerUnderVendor = vendorDirectory.getChild(repo.getMarkerFileName());
-        Path repoUnderVendor = vendorDirectory.getRelative(repo.getName());
-
-        // 1. Clean up existing marker file and repo vendor directory
-        markerUnderVendor.delete();
-        repoUnderVendor.deleteTree();
-        repoUnderVendor.createDirectory();
-
-        // 2. Copy over the repo source.
-        FileSystemUtils.copyTreesBelow(
-            externalRepoRoot.getRelative(repo.getName()), repoUnderVendor, Symlinks.NOFOLLOW);
+      Path repoUnderExternal = externalRepoRoot.getChild(repo.getName());
+      Path repoUnderVendor = vendorDirectory.getRelative(repo.getName());
+      // This could happen when running the vendor command twice without changing anything.
+      if (repoUnderExternal.isSymbolicLink() && repoUnderExternal.resolveSymbolicLinks().compareTo(repoUnderVendor) == 0) {
+        continue;
+      }
 
-        // 3. Copy the marker file atomically
-        Path tMarker = vendorDirectory.getChild(repo.getMarkerFileName() + ".tmp");
-        FileSystemUtils.copyFile(externalRepoRoot.getChild(repo.getMarkerFileName()), tMarker);
-        tMarker.renameTo(markerUnderVendor);
+      // At this point, the repo should exist under external dir, but check if the vendor src is already up-to-date.
+      Path markerUnderExternal = externalRepoRoot.getChild(repo.getMarkerFileName());
+      Path markerUnderVendor = vendorDirectory.getChild(repo.getMarkerFileName());
+      if (isRepoUpToDate(markerUnderVendor, markerUnderExternal)) {
+        continue;
       }
+
+      // Actually vendor the repo:
+      // 1. Clean up existing marker file and vendor dir.
+      markerUnderVendor.delete();
+      repoUnderVendor.deleteTree();
+      repoUnderVendor.createDirectory();
+      // 2. Move the marker file to a temporary one under vendor dir.
+      Path tMarker = vendorDirectory.getChild(repo.getMarkerFileName() + ".tmp");
+      FileSystemUtils.moveFile(markerUnderExternal, tMarker);
+      // 3. Move the external repo to vendor dir. It's fine if this step fails or is interrupted, because the marker
+      // file under external is gone anyway.
+      FileSystemUtils.moveTreesBelow(repoUnderExternal, repoUnderVendor);
```

Comment:
How does this behave if a repo symlinks files from another repo and one is vendored while the other is not? It looks like it may be necessary to follow relative symlinks but not absolute symlinks.

Reply:
The `moveTreesBelow` doesn't follow any symlinks. Judging from the code [here](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryContext.java;l=214-215), it's actually impossible to create relative symlink with the `ctx.symlink` API.

I tested with
```
  ctx.symlink("/tmp/foo", "path_abs")
  ctx.symlink("data", "path_rel")
  ctx.symlink(ctx.path(Label("@bar//:data")), "path_bar")
  ```
and it resulted

```
path_abs@ -> /tmp/foo
path_bar@ -> /private/var/tmp/_bazel_pcloudy/d278f827a729facdbfb1ff0fc0002042/external/_main~ext~bar/data
path_rel@ -> /private/var/tmp/_bazel_pcloudy/d278f827a729facdbfb1ff0fc0002042/external/_main~ext~foo/data
```
in both external and vendor dir.

This is fine if only foo is vendored, since eventually `<output_base>/external/_main~ext~bar` would exist and point to the right location. However, I noticed there is problem if output base is changed after vendoring.

Reply:
While this is the current behavior, wouldn't we have to change it so that symlinks in vendored repos do not contain absolute paths? I think there was another issue about this files recently.

Reply:
> While this is the current behavior, wouldn't we have to change it so that symlinks in vendored repos do not contain absolute paths? I think there was another issue about this filed recently.

https://github.com/bazelbuild/bazel/issues/22303, probably

Reply:
> To deal with potential output base, maybe we could
> 
> 1. create a symlink pointing to the external repo root under the vendor dir
> 2. Rewrite all symlinks pointing some path under external repo root to a relative path to the symlink created in 1.

this is quite clever! but what will version-control systems do with this special symlink? Usually people put `bazel-*` symlinks in the workspace root in .gitignore, so presumably this new special symlink will also need to be ignored? And the symlink is generated on demand if it's not there, etc.? (I agree that this should be done in a separate PR)

Either way, some sort of symlink rewriting will need to happen, and we'll probably need to do something similar for the true repo cache.

Reply:
>  so presumably this new special symlink will also need to be ignored? And the symlink is generated on demand if it's not there, etc.? (I agree that this should be done in a separate PR)

Yes, I also think it should be gitignored since it's machine specific. And we can just always create the symlink since it's quite cheap to keep the code simpler.